src/Simplify.cpp

```diff
@@ -729,7 +741,19 @@ class Simplify : public IRMutator {
         } else if (broadcast_a && broadcast_b) {
             // Broadcast + Broadcast
             expr = Broadcast::make(mutate(broadcast_a->value - broadcast_b->value),
-                                 broadcast_a->lanes);
+                                   broadcast_a->lanes);
+        } else if (cast_a && cast_b &&
+                   cast_a->value.type() == cast_b->value.type() &&
+                   cast_a->type.is_int() &&
+                   cast_b->type.is_int() &&
+                   cast_a->value.type().is_int() &&
+                   cast_b->value.type().is_int() &&
+                   no_overflow(cast_a->type) &&
+                   no_overflow(cast_b->type) &&
+                   no_overflow(cast_a->value.type()) &&
+                   no_overflow(cast_b->value.type())) {
+            // cast(t, a) - cast(t, b) -> cast(t, a - b)
```

Comment:
Simplify rules are not allowed to change the bits of a result for arguments that are valid. I believe this one does. (It would perhaps be a bit clearer if the "cast_a->value.type() == cast_b->value.type()" were done in one if statement and then a "src_type" variable was introduced to indicate there is a single starting type for both sides.) 

 Likely the issue here is that we're trying for a design where addresses are guaranteed to follow certain rules such that overflow does not occur. (Which is what no_overflow is saying.) But a) there is nothing that really solidly separates addresses and user data values in the IR so we do it by a heuristic on type, and b) addresses get mixed with coordinates and if we really push on coordinates large enough, the model breaks.

no_overflow never returns true for vectors if I'm reading the code correctly. I would expect this rule to never fire for vectors, which may or may not be an issue. (I.e. I think our heuristic is that only base addresses are in the no_overflow camp and the ramps applied to them are not.) Mentioning it per Andrews comment on the downsize cast case.